text/0000-procedural-vtables.md

```diff
@@ -0,0 +1,338 @@
+/// default trait objects. Your own trait objects can use any metadata and
+/// thus "vtable" layout that they want.
+unsafe impl Unsized for dyn MyTrait {
+    type WidePointer = Pointer;
+    fn size_of(ptr: Pointer) -> usize {
+        ptr.vtable.size
+    }
+    fn align_of(ptr: Pointer) -> usize {
+        ptr.vtable.align
+    }
+}
+
+impl Drop for dyn MyTrait {
+    fn drop(&mut self) {
+        // using a dummy concrete type for `Pointer`
+        let ptr = transmute::<&mut dyn Trait, Pointer<()>>(self);
+        unsafe {
+            let drop = ptr.vtable.drop;
+            drop(&raw mut ptr.ptr)
+        }
+    }
+}
+
+const fn default_vtable<T: MyTrait>() -> VTable<T> {
+    // `VTable` is a `#[repr(C)]` type with fields at the appropriate
+    // places.
+    VTable {
+        size: std::mem::size_of::<T>(),
+        align: std::mem::size_of::<T>(),
+        drop: <T as Drop>::drop,
+        some_fn: fn (&mut T, u32) -> i32,
+    }
+}
+```
+
+Now, if you want to implement a fancier vtable, this RFC enables you to do that.
+
+## Null terminated strings (std::ffi::CStr)
+
+This is how I see all extern types being handled.
+There can be no impls of `CStr` for any type, because the `Unsize`
+trait impl is missing. See the future extension section at the end of this RFC for
+ideas that could support `CString` -> `CStr` unsizing by allowing `CString` to implement
+`CStr` instead of having a `Deref` impl that converts.
+
+```rust
+pub trait CStr {}
+
+impl CustomUnsized<CStrPtr> for dyn CStr {
+    type WidePointer = *mut u8;
+    fn size_of(ptr: *mut u8) -> usize {
+        unsafe { strlen(ptr) }
+    }
+    fn align_of(_: *mut u8) -> usize {
+        1
+    }
+}
+```
+
+## `[T]` as sugar for a `Slice` trait
+
+We could remove `[T]` (and even `str`) from the language and just make it desugar to
+a `std::slice::Slice` (or `StrSlice`) trait.
+
+```rust
+pub trait Slice<T> {}
+
+#[repr(C)]
+struct SlicePtr<T> {
+    ptr: *mut T,
+    len: usize,
+}
+
+// This impl must be in the `vec` module, to give it access to the `vec`
+// internals instead of going through `&mut Vec<T>` or `&Vec<T>`.
+impl<T> CustomUnsize<SlicePtr<T>> for Vec<T> {
+    fn unsize<const owned: bool>(ptr: *mut Vec<T>) -> SlicePtr<T> {
+        SlicePtr {
+            ptr: vec.data,
+            len: vec.len,
+        }
+    }
+}
+
+impl<T> CustomUnsized for dyn Slice<T> {
+    type WidePointer = SlicePtr<T>;
+    fn size_of(ptr: SlicePtr<T>) -> usize {
+        ptr.len
+    }
+    fn align_of(_: SlicePtr<T>) -> usize {
+        std::mem::align_of::<T>()
+    }
+}
+
+impl Drop for dyn Slice<T> {
+    fn drop(&mut self) {
+        let wide_ptr = transmute::<&mut dyn Slice<T>, SlicePtr<T>>(self);
+        let mut data_ptr = wide_ptr.ptr;
+        for i in 0..wide_ptr.len {
+            std::ptr::drop_in_place(data_ptr);
+            data_ptr = data_ptr.offset(1);
+        }
+    }
+}
+```
+
+## C++ like vtables
+
+Most of the boilerplate is the same as with regular vtables.
```

Comment:
Might be worth clarifying how this could coexist with the existing C++ vtables RFC.